mwclient/client.py

```diff
@@ -521,7 +568,7 @@ def raw_call(self, script, data, files=None, retry_on_error=True, http_method='P
                 if not retry_on_error:
                     raise
                 log.warning('Connection error. Retrying in a moment.')
-                toraise = err
+                toraise = err  # type: ignore[assignment]
```

Comment:
could we make this `Union[BaseException, str]`, since it has to be something that can be raised or the string `"stream"`?

Reply:
Technically it should be `Optional[Union[requests.RequestException, Literal['stream']]]` I guess, which would make the if/else type-check cleanly, but `Litaral` requires Python 3.8+. For now we need to go with something like this I guess:

```diff
diff --git a/mwclient/client.py b/mwclient/client.py
index a1e0cde..5593a32 100644
--- a/mwclient/client.py
+++ b/mwclient/client.py
@@ -527,7 +527,7 @@ class Site:
         url = f'{scheme}://{host}{self.path}{script}{self.ext}'
 
         while True:
-            toraise = None
+            toraise = None  # type: Optional[Union[requests.RequestException, str]]
             wait_time = 0
             args = {'files': files, 'headers': headers}  # type: Dict[str, Any]
             for k, v in self.requests.items():
@@ -568,7 +568,7 @@ class Site:
                 if not retry_on_error:
                     raise
                 log.warning('Connection error. Retrying in a moment.')
-                toraise = err  # type: ignore[assignment]
+                toraise = err
                 # proceed to the sleep
 
             # all retry paths come here
@@ -577,8 +577,8 @@ class Site:
             except errors.MaximumRetriesExceeded:
                 if toraise == "stream":
                     stream.raise_for_status()
-                elif toraise:
-                    raise toraise  # type: ignore[misc]
+                elif toraise and isinstance(toraise, Exception):
+                    raise toraise
                 else:
                     raise
```

Reply:
Alternatively we could require typing-extensions on python versions <3.8:

```diff
diff --git a/mwclient/client.py b/mwclient/client.py
index 18e3b09..991759f 100644
--- a/mwclient/client.py
+++ b/mwclient/client.py
@@ -1,11 +1,17 @@
 import io
 import json
 import logging
+import sys
 import warnings
 from collections import OrderedDict
 from typing import Optional, Callable, Union, Mapping, Any, MutableMapping, List, Dict, \
     Tuple, cast, Iterable, BinaryIO, TextIO
 
+if sys.version_info < (3, 8):
+    from typing_extensions import Literal  # noqa: F401
+else:
+    from typing import Literal  # noqa: F401
+
 import requests
 from requests.auth import HTTPBasicAuth, AuthBase
 from requests_oauthlib import OAuth1
@@ -527,7 +533,7 @@ class Site:
         url = f'{scheme}://{host}{self.path}{script}{self.ext}'
 
         while True:
-            toraise = None
+            toraise = None  # type: Union[Literal['stream'], requests.exceptions.RequestException, None]  # noqa: E501
             wait_time = 0
             args = {'files': files, 'headers': headers}  # type: Dict[str, Any]
             for k, v in self.requests.items():
@@ -568,7 +574,7 @@ class Site:
                 if not retry_on_error:
                     raise
                 log.warning('Connection error. Retrying in a moment.')
-                toraise = err  # type: ignore[assignment]
+                toraise = err
                 # proceed to the sleep
 
             # all retry paths come here
@@ -578,7 +584,7 @@ class Site:
                 if toraise == "stream":
                     stream.raise_for_status()
                 elif toraise:
-                    raise toraise  # type: ignore[misc]
+                    raise toraise
                 else:
                     raise
 
diff --git a/setup.py b/setup.py
index 9420efe..a92fdbd 100644
--- a/setup.py
+++ b/setup.py
@@ -30,7 +30,10 @@ setup(name='mwclient',
       url='https://github.com/mwclient/mwclient',
       license='MIT',
       packages=['mwclient'],
-      install_requires=['requests-oauthlib'],
+      install_requires=[
+          'requests-oauthlib',
+          'typing-extensions; python_version < "3.8"',
+      ],
       extras_require={
           'testing': ['pytest', 'pytest-cov',
                       'responses>=0.3.0', 'responses!=0.6.0', 'setuptools'],
```

Reply:
I think I prefer the former, WDYT?

Reply:
Yep, let's go with that for now. Fixed in 1baa435.